File: src/class041/C3_SameMod.py
import random

# This class is about the modulo property of addition, subtraction, and multiplication
# It doesn't cover division as division requires inverse modulo, which will be covered later
class SameMod:

    # For testing purposes
    @staticmethod
    def random():
        return random.randint(0, 2**63-1)

    # Calculate ((a + b) * (c - d) + (a * c - b * d)) % mod with non-negative result
    @staticmethod
    def f1(a, b, c, d, mod):
        o1 = a % mod
        o2 = b % mod
        o3 = c % mod
        o4 = d % mod
        o5 = (o1 + o2) % mod
        o6 = (o3 - o4 + mod) % mod
        o7 = (o1 * o3) % mod
        o8 = (o2 * o4) % mod
        o9 = (o5 * o6) % mod
        o10 = (o7 - o8 + mod) % mod
        return (o9 + o10) % mod

    @staticmethod
    def f2(a, b, c, d, mod):
        # Same logic but with different calculation style
        o1, o2, o3, o4 = a % mod, b % mod, c % mod, d % mod
        o5 = (o1 + o2) % mod
        o6 = (o3 - o4 + mod) % mod
        o7 = (o1 * o3) % mod
        o8 = (o2 * o4) % mod
        o9 = (o5 * o6) % mod
        o10 = (o7 - o8 + mod) % mod
        return (o9 + o10) % mod

    @staticmethod
    def main():
        print("Testing starts")
        testTime = 100000
        mod = 1000000007
        for _ in range(testTime):
            a, b, c, d = SameMod.random(), SameMod.random(), SameMod.random(), SameMod.random()
            if SameMod.f1(a, b, c, d, mod) != SameMod.f2(a, b, c, d, mod):
                print("Error!")
        print("Testing ends")

        print("===")
        a, b, c, d = SameMod.random(), SameMod.random(), SameMod.random(), SameMod.random()
        print(f"a: {a}")
        print(f"b: {b}")
        print(f"c: {c}")
        print(f"d: {d}")
        print("===")
        print(f"f1: {SameMod.f1(a, b, c, d, mod)}")
        print(f"f2: {SameMod.f2(a, b, c, d, mod)}")

if __name__ == '__main__':
    SameMod.main()

File: src/class041/C3_SameMod.cpp
#include <iostream>
#include <random>

// This class is about the modulo property of addition, subtraction, and multiplication
// It doesn't cover division as division requires inverse modulo, which will be covered later
class SameMod {
public:
    // For testing purposes
    static long long random() {
        std::random_device rd;
        std::mt19937_64 eng(rd());
        std::uniform_int_distribution<long long> distr;
        return distr(eng);
    }

    // Calculate ((a + b) * (c - d) + (a * c - b * d)) % mod with non-negative result
    static int f1(long long a, long long b, long long c, long long d, int mod) {
        long long o1 = a % mod;
        long long o2 = b % mod;
        long long o3 = c % mod;
        long long o4 = d % mod;
        long long o5 = (o1 + o2) % mod;
        long long o6 = (o3 - o4 + mod) % mod;
        long long o7 = (o1 * o3) % mod;
        long long o8 = (o2 * o4) % mod;
        long long o9 = (o5 * o6) % mod;
        long long o10 = (o7 - o8 + mod) % mod;
        return (int)((o9 + o10) % mod);
    }

    static void test() {
        std::cout << "Testing starts" << std::endl;
        int testTime = 100000;
        int mod = 1000000007;
        for (int i = 0; i < testTime; ++i) {
            long long a = random();
            long long b = random();
            long long c = random();
            long long d = random();
            if (f1(a, b, c, d, mod) != f1(a, b, c, d, mod)) {
                std::cout << "Error!" << std::endl;
            }
        }
        std::cout << "Testing ends" << std::endl;

        std::cout << "===" << std::endl;
        long long a = random();
        long long b = random();
        long long c = random();
        long long d = random();
        std::cout << "a: " << a << std::endl;
        std::cout << "b: " << b << std::endl;
        std::cout << "c: " << c << std::endl;
        std::cout << "d: " << d << std::endl;
        std::cout << "===" << std::endl;
        std::cout << "f1: " << f1(a, b, c, d, mod) << std::endl;
    }
};

int main() {
    SameMod::test();
    return 0;
}
